### ISSUE TITLE
False type mismatch between function and `Protocol` when specifying `__name__` in the `Protocol`

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#2179]
**Describe the bug**
A type mismatch error is reported when passing a function as an argument to another function that expects a specific `Protocol` as the argument, when the function correctly implements the `Protocol`.

More specifically, when specifying `__name__: str` in the `Protocol`, pyright will report that the function is missing attributes:  `__name__` and `__call__`.

**To Reproduce**
Create snippet as such:

```py
from typing import Protocol


class SomeFunc(Protocol):
    __name__: str

    def __call__(self) -> str:
        ...


def other_func(f: SomeFunc):
    print(f.__name__)


@other_func
def some_func() -> str:
    ...
```

Pyright will now report the error:
```
error: Argument of type "() -> str" cannot be assigned to parameter "f" of type "SomeFunc" in function "other_func"
   "__name__" is not present
   "__call__" is not present
```

However this is not consistent with python:
```
>>> def some_func():
...   ...
... 
>>> some_func.__name__
'some_func'
>>> some_func.__call__
<method-wrapper '__call__' of function object at 0x1021b61f0>
```

**Expected behavior**
I would expect pyright not to report an error in this case. Changing the type annotation to use `Callable[[], str]` in `other_func` results in no type errors:

```py
from typing import Callable


def other_func(f: Callable[[], str]):
    print(f.__name__)


@other_func
def some_func() -> str:
    ...

```
So I would expect the same to happen when using a `Protocol`.

This seems to have been a recent change, as I did not get any errors on `1.1.159`.

Tested on pyright version `1.1.161`

**Comment by erictraut** [comment]
I think the current behavior is correct. You are defining a protocol class that includes a `__call__` and a `__name__`. I suspect that you intend for this to be used as a [callback protocol](https://www.python.org/dev/peps/pep-0544/#callback-protocols), but callback protocols are not allowed to include any symbols other than `__call__`.

To further bolster my confidence that the current behavior is correct, it matches the behavior of mypy in this case.

If this previously worked in pyright, then it was an unintended behavior and a bug.

**Comment by kennipj** [comment]
That seems like an odd limitation to me. Could you elaborate on the decision to disallow this behaviour?

As far as I can see, in PEP544 there is no explicit mentioning of callback protocols being restricted to only having `__call__` members:
> Protocols can be used to define flexible callback types that are hard (or even impossible) to express using the Callable[...] syntax specified by PEP 484, such as variadic, overloaded, and complex generic callbacks. They can be defined as protocols *with* a `__call__` member:

From that I gather that the only requirement is that `__call__` member is present in the `Protocol`, and that there are no other apparent limitations.

Additionally, after the code block it is mentioned that:
> Callback protocols and Callable[...] types can be used interchangeably.

But if this is to be true, then this leads me to a different issue in pyright, as these are not interchangable. See the following example:
```py
from typing import Protocol


class SomeFunc(Protocol):
    def __call__(self) -> str:
        ...


def some_func(f: SomeFunc):
    f.__name__ # Cannot access member "__name__" for type "SomeFunc"
```

Whereas with `Callable` we get:

```py
from typing import Callable


def some_func(f: Callable[[], str]):
    f.__name__ # No error
```

I will gladly create another issue report for this example if you think this not as designed.

**Comment by erictraut** [comment]
This is intended behavior as well. There is a symbol defined in builtins.pyi called `function`, and it defines the attributes that are legal for a `Callable` type. The `__name__` attribute is included here.

```python
class function:
    __name__: str
    __module__: str
    __code__: CodeType
    __qualname__: str
    __annotations__: dict[str, Any]
```

A callback protocol does not have a `__name__` attribute.

Once again, mypy emits the same errors as pyright in this case.

**Comment by kennipj** [comment]
The mypy repo currently has an open issue with the same issue I'm describing here, labeled as a bug: https://github.com/python/mypy/issues/10403.

So I'm not sure the current mypy behaviour is intended either.

**Comment by povilasb** [comment]
1. Just to highlight again, Kenni was not trying to use a `Callable` but rather his own `NamedCallable` protocol:
```
class NamedCallable(Protocol):
    __name__: str

    def __call__(self) -> str:
        ...
```

2. Now it seems to me that the actual problem is why the `function` is interpreted as a `Callable`:
```
def func() -> str:
    return ""
reveal_type(func)
"Callable[[], str]"
```
even though there is a `function` type definition:
```
class function:
    __name__: str
    __module__: str
    __code__: CodeType
    __qualname__: str
    __annotations__: dict[str, Any]
```

Despite the fact that `mypy` works the same way :)

**Comment by erictraut** [comment]
After consulting with the broader Python type community and after much internal discussion, we've decided to change our interpretation of "callback protocol" and treat any protocol with a `__call__` method as a callback protocol even if it includes additional attributes. With this change, the sample at the top of this issue now type checks without errors.

**Comment by erictraut** [comment]
This is included in Pyright 1.1.163, which I just published. It will also be in the next release of Pylance.
